From 53dab6b655c62d5686f05519e5a5f04597373f1d Mon Sep 17 00:00:00 2001
From: Serguei Katkov <serguei.i.katkov@intel.com>
Date: Fri, 15 Feb 2013 14:39:40 +0700
Subject: Dalvik: GC scanning grey objects should match the GC asserts

BZ: 65653

Scanning grey objects is not accurate in terms of to be correspondent to assert statements.
The patch fixes several issues:
1) Utility function dvmCardFromAddr requires the real address matching the bounds
while dvmHeapSourceGetLimit() return the first address greater than bound.
The patch uses dvmHeapSourceGetLimit() - 1 address getting the last card
and then adds 1 to get the card limit.
2) assert(limit <= &h->cardTableBase[h->cardTableLength]) does not take into account
that card table is shifted by offset. Patch adds offset.
3) It possible that scanDirtyCards goes outside of the bound while
dvmHeapBitmapIsObjectBitSet used in nextGrayObject requires the address in bounds.
Patch fixes this issue by not allowing to go outside of bounds.

Category: device-enablement
Domain: AOSP-Dalvik-GC
Origin: internal
Upstream-Candidate: yes

Change-Id: I8fa86e36c195787ca0b9444edbe1c0839d6d1a88
Orig-MCG-Change-Id: Id2386a87b931bc2c21c31a2c3008b679c18204a5
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/alloc/MarkSweep.cpp |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/vm/alloc/MarkSweep.cpp b/vm/alloc/MarkSweep.cpp
index 8cc062a..1450e6f 100644
--- a/vm/alloc/MarkSweep.cpp
+++ b/vm/alloc/MarkSweep.cpp
@@ -558,8 +558,9 @@ const u1 *scanDirtyCards(const u1 *start, const u1 *end,
         if (*card != GC_CARD_DIRTY) {
             return card;
         }
-        const u1 *ptr = prevAddr ? prevAddr : (u1*)dvmAddrFromCard(card);
-        const u1 *limit = ptr + GC_CARD_SIZE;
+        const u1 *cardPtr = (u1*)dvmAddrFromCard(card);
+        const u1 *ptr = prevAddr ? prevAddr : cardPtr;
+        const u1 *limit = cardPtr + GC_CARD_SIZE;
         while (ptr < limit) {
             Object *obj = nextGrayObject(ptr, limit, markBits);
             if (obj == NULL) {
@@ -568,7 +569,7 @@ const u1 *scanDirtyCards(const u1 *start, const u1 *end,
             scanObject(obj, ctx);
             ptr = (u1*)obj + ALIGN_UP(objectSize(obj), HB_OBJECT_ALIGNMENT);
         }
-        if (ptr < limit) {
+        if (ptr <= limit) {
             /* Ended within the current card, advance to the next card. */
             ++card;
             prevAddr = NULL;
-- 
1.7.4.1

